Refactoring of ics_cobraOps

jagracar opened the pull request.

wtgee reviewed (approved):
LGTM w.r.t. any dependencies on shared tools. There might be some changes to `ics_cobraCharmer` coming up but it doesn't look like this should affect it.

The `demo.py` and `conftest.py` contain some hard-coded local paths that will make it harder for other people to test.

FYI, we'll probably be switching the actors to `pyproject.toml` instead of `setup.py` but that will likely happen in the future and shouldn't affect how things are done.

mreineck commented:
I have switched the package to `pyproject.toml` now, but I'm not a real expert on this. Please have a good look at this; if I broke something, we can just revert.

monodera commented:
Hi, I'm trying to make the target uploader (https://github.com/Subaru-PFS/spt_target_uploader) to be compatible to this refactoring. Previously, it was run with `Bench(layout="full")`, but this option is no longer available, and I need to set `cobraCoach` and `blackDotsCalibrationProduct`. These parameters seem to require `pfs_instdata` installed, but `pfs_instdata` is not pip-installable. 

Moreover, `cobraCoach` import makes the cascade of dependencies which is not easy to fix (first, `ics_cobraChermer` requres `sep`, but not not listed in any of `pyprojects.toml`). Then, import fails as follows

```
  File "/Users/monodera/Dropbox/NAOJ/PFS/spt_target_uploader/src/pfs_target_uploader/utils/ppp.py", line 34, in <module>
    from ics.cobraCharmer.cobraCoach.cobraCoach import CobraCoach
  File "/Users/monodera/Dropbox/NAOJ/PFS/spt_target_uploader/.venv/src/ics-cobraCharmer/python/ics/cobraCharmer/cobraCoach/cobraCoach.py", line 12, in <module>
    from ics.cobraCharmer.cobraCoach.mcs import camera
  File "/Users/monodera/Dropbox/NAOJ/PFS/spt_target_uploader/.venv/src/ics-cobraCharmer/python/ics/cobraCharmer/cobraCoach/mcs/camera.py", line 13, in <module>
    from ics.fpsActor import najaVenator
ModuleNotFoundError: No module named 'ics.fpsActor'
```

Then, `ics_fpsActor` is not pip-installable, and `najaVenator` is at the top, but not actually used in the `camera` module (if I remember correctly). 

Could you please let me know how to adopt this refactoring in a nice way?

jagracar commented:
Unfortunately the only way i found was to clone all the pfs required projects and added their python directories to my python path.

For the python dependencies, you will have to install them one by one.

I recently changed my laptop and these are the deb packages that i installed (some might not be needed for PFS):

```
  sudo apt install ipython3
  sudo apt install python3-pip
  sudo apt install python3-numpy
  sudo apt install python3-matplotlib
  sudo apt install python3-pandas
  sudo apt install python3-scipy
  sudo apt install python3-shapely
  sudo apt install python3-astropy
  sudo apt install python3-opencv
  sudo apt install python3-sep
  sudo apt install python3-astroplan
  sudo apt install python3-sqlalchemy
  sudo apt install python3-psycopg2

```

These are the PFS projects that i cloned locally:

```
  git clone https://github.com/Subaru-PFS/ics_cobraOps.git
  git clone https://github.com/Subaru-PFS/ics_cobraCharmer
  git clone https://github.com/Subaru-PFS/ets_pointing
  git clone https://github.com/Subaru-PFS/ets_fiberalloc
  git clone https://github.com/Subaru-PFS/pfs_utils
  git clone https://github.com/Subaru-PFS/pfs_instdata
  git clone https://github.com/Subaru-PFS/spt_operational_database
  git clone https://github.com/Subaru-PFS/ics_fpsActor
```

And this is how i added the paths in my .bashrc file:

`export PYTHONPATH="${PYTHONPATH}:/home/jgracia/github/ics_cobraOps/python:/home/jgracia/github/ics_cobraCharmer/python:/home/jgracia/github/ics_fpsActor/python/:/home/jgracia/github/pfs_utils/python:/home/jgracia/github/spt_operational_database/python"
`

mreineck commented:
I'm asking because this might lead to collisions between different versions of the same package rather quickly, in contrast to being able to specify version ranges.

wtgee commented:
> I'm asking because this might lead to collisions between different versions of the same package rather quickly, in contrast to being able to specify version ranges.

Unfortunately that is correct, you cannot specify a range or minimum, it must be a single tag, branch, or commit.

mreineck commented:
Thanks for the clarification! I'll use this for netflow in the future.

jagracar commented:
I will update the cobraOps file to point to the correct version of cobraCharmer

jagracar commented:
Here is the updated file
https://github.com/Subaru-PFS/ics_cobraOps/blob/master/pyproject.toml

Btw, i realized that when i created the v2.0.0 tag i forgot to increase the version here:
https://github.com/Subaru-PFS/ics_cobraOps/blob/v2.0.0/python/ics/cobraOps/_version.py

Would that create trouble?

I can create  a new tag if it does.

wtgee commented:
> Do we need to add this too?

@jagracar `hatch` is a tool that allows you to manage the package more easily and if using `hatch` you do need that option in order to have github dependencies. You don't need to use hatch but it also doesn't hurt to have that line in your pyproject.toml, it will basically just be ignored unless you are using hatch.

> Here is the updated file

Package names should have underscores instead of dots, so `ics_cobraCharmer` rather than `ics.cobraCharmer`.  This is in the dependencies section as well as the actual name (line 2).

So the package name is `ics_cobraCharmer`, the python import is `ics.cobraCharmer`, and the file path is `ics/cobraCharmer`.

⚠️ 

One other issue that we need to remove the `__init__.py` file in the root of the package (here: https://github.com/Subaru-PFS/ics_cobraOps/blob/master/python/ics/__init__.py).

This has to do with the "namespace" of our packages. Here our namespace is `ics.` (for the pipeline the namespace is `pfs.`...really our ICS stuff should be `pfs.ics.` but that's too much of a change). 

Because of how python imports work and because our `ics` packages exist in different repositories we *cannot* have the `__init__.py` file in the root `python/ics` directory and that has to be true of *all* `ics.` packages that are installed together.  Essentially if any single repo has a `__init__.py` file in `python/ics/` then when you do a `pip install` on that package it will assume everything in that namespace is in that directory.  
The contents of the current `__init__.py` in `ics_cobraOps` are outdated so that file can be safely removed.

This is the more annoying and difficult change because it requires every similarly namespaced package to be correct although it's already been removed for `ics_cobraCharmer` and `pfs_instdata`, which are your only PFS dependencies.

I've also removed those files for `ics.utils` and `pfs.utils` and a few of the other ICS but not all of them. After this current run I will probably go through systematically and remove them but if you are making changes now you should go ahead and remove it. (This is [PEP 420](https://peps.python.org/pep-0420/) for reference)

One last note, you have a typo in `psf.instdata` on line 15 (`psf` instead of `pfs`)
